File: tools/generate_sitemap.py
# =============================================================================
#   THE JESUS WEBSITE — SEO SITEMAP GENERATOR
#   File:    tools/generate_sitemap.py
#   Version: 1.2.0
#   Purpose: Queries SQLite to dynamically build an XML sitemap for search engines.
#            Updated for URL slug restructure — uses clean paths instead of
#            raw .html filenames, and path-based record slugs.
# =============================================================================

import datetime
import logging
import os
import sqlite3
import sys

# Ensure package context is recognized when running directly from CLI
sys.path.append(os.path.dirname(os.path.dirname(os.path.abspath(__file__))))
from backend.middleware.logger_setup import setup_logger

# Initialize central logging to /logs
logger = setup_logger(__file__)

ROOT_DIR = os.path.dirname(os.path.dirname(os.path.abspath(__file__)))
DB_PATH = os.path.join(ROOT_DIR, "database", "database.sqlite")
SITEMAP_OUTPUT_PATH = os.path.join(ROOT_DIR, "frontend", "pages", "sitemap.xml")
BASE_URL = "https://www.thejesuswebsite.com"


def generate_sitemap():
    """
    Trigger:  Run directly via CLI or via build.py.
    Function: Reads all records slugs from SQLite and writes a standards-compliant
              XML sitemap to frontend/pages/sitemap.xml.
    Output:   sitemap.xml at SITEMAP_OUTPUT_PATH, combining static routes,
              record deep-dive URLs, and blog post URLs.
    """
    logger.info("Starting internal XML Sitemap Generator...")

    if not os.path.exists(DB_PATH):
        logger.error("Database not found. Generating static sitemap only.")
        records = []
        blog_posts = []
    else:
        try:
            conn = sqlite3.connect(DB_PATH)
            conn.row_factory = sqlite3.Row
            cursor = conn.cursor()

            # Fetch all records with slugs for record deep-dive pages
            cursor.execute(
                "SELECT slug, updated_at FROM records WHERE slug IS NOT NULL"
            )
            records = [dict(row) for row in cursor.fetchall()]

            # Fetch records that have blogpost content for individual blog post pages
            # Uses the `blogposts` TEXT column on the records table (JSON blob).
            # Gracefully handles the case where the column does not yet exist.
            try:
                cursor.execute(
                    "SELECT slug, updated_at FROM records WHERE blogposts IS NOT NULL AND blogposts != ''"
                )
                blog_posts = [dict(row) for row in cursor.fetchall()]
            except sqlite3.OperationalError:
                logger.info("blogposts column not found — skipping blog post URLs.")
                blog_posts = []

            conn.close()
        except sqlite3.OperationalError as e:
            logger.error(f"Cannot access records table: {str(e)}")
            records = []
            blog_posts = []

    # Boilerplate XML setup
    xml_content = '<?xml version="1.0" encoding="UTF-8"?>\n'
    xml_content += '<urlset xmlns="http://www.sitemaps.org/schemas/sitemap/0.9">\n'

    # 1. Static routes — clean slugs replacing legacy `.html` paths
    #    Includes critical landing pages from the slug map.
    static_routes = [
        ("/", "1.0", "daily"),
        ("/about", "0.8", "monthly"),
        ("/context", "0.9", "weekly"),
        ("/debate", "0.9", "weekly"),
        ("/resources", "0.9", "weekly"),
        ("/records", "0.9", "weekly"),
        ("/evidence", "0.8", "monthly"),
        ("/timeline", "0.8", "monthly"),
        ("/maps", "0.8", "monthly"),
        ("/news", "0.7", "weekly"),
        ("/news/feed", "0.7", "weekly"),
        ("/blog", "0.7", "weekly"),
        ("/blog/post", "0.6", "monthly"),
        ("/context/essay", "0.7", "monthly"),
        ("/debate/academic-challenges", "0.7", "monthly"),
        ("/debate/popular-challenges", "0.7", "monthly"),
        ("/debate/wikipedia-articles", "0.7", "monthly"),
        ("/debate/historiography", "0.7", "monthly"),
        ("/debate/response", "0.6", "monthly"),
        ("/resources/events", "0.6", "monthly"),
        ("/resources/external-witnesses", "0.6", "monthly"),
        ("/resources/internal-witnesses", "0.6", "monthly"),
        ("/resources/manuscripts", "0.6", "monthly"),
        ("/resources/miracles", "0.6", "monthly"),
        ("/resources/ot-verses", "0.6", "monthly"),
        ("/resources/objects", "0.6", "monthly"),
        ("/resources/people", "0.6", "monthly"),
        ("/resources/sermons-and-sayings", "0.6", "monthly"),
        ("/resources/sites", "0.6", "monthly"),
        ("/resources/sources", "0.6", "monthly"),
        ("/resources/world-events", "0.6", "monthly"),
        ("/maps/roman-empire", "0.6", "monthly"),
        ("/maps/galilee", "0.6", "monthly"),
        ("/maps/jerusalem", "0.6", "monthly"),
        ("/maps/judea", "0.6", "monthly"),
        ("/maps/levant", "0.6", "monthly"),
    ]

    today = datetime.datetime.now(datetime.UTC).strftime("%Y-%m-%d")
    for route, priority, freq in static_routes:
        xml_content += f"  <url>\n"
        xml_content += f"    <loc>{BASE_URL}{route}</loc>\n"
        xml_content += f"    <lastmod>{today}</lastmod>\n"
        xml_content += f"    <changefreq>{freq}</changefreq>\n"
        xml_content += f"    <priority>{priority}</priority>\n"
        xml_content += f"  </url>\n"

    # 2. Record deep-dive pages — path-based slugs (e.g. /record/jesus-baptism)
    #    Nginx rewrites /record/{slug} internally to record.html?slug={slug}
    for record in records:
        slug = record.get("slug")
        raw_updated = record.get("updated_at")
        last_mod = raw_updated.split("T")[0] if raw_updated else today

        xml_content += f"  <url>\n"
        xml_content += f"    <loc>{BASE_URL}/record/{slug}</loc>\n"
        xml_content += f"    <lastmod>{last_mod}</lastmod>\n"
        xml_content += f"    <changefreq>monthly</changefreq>\n"
        xml_content += f"    <priority>0.7</priority>\n"
        xml_content += f"  </url>\n"

    # 3. Blog post pages — uses ?id= query param (same convention as /context/essay)
    #    These entries appear in the sitemap as soon as the blogposts column
    #    is populated on any record.
    for post in blog_posts:
        slug = post.get("slug")
        raw_updated = post.get("updated_at")
        last_mod = raw_updated.split("T")[0] if raw_updated else today

        xml_content += f"  <url>\n"
        xml_content += f"    <loc>{BASE_URL}/blog/post?id={slug}</loc>\n"
        xml_content += f"    <lastmod>{last_mod}</lastmod>\n"
        xml_content += f"    <changefreq>monthly</changefreq>\n"
        xml_content += f"    <priority>0.6</priority>\n"
        xml_content += f"  </url>\n"

    xml_content += "</urlset>\n"

    # Ensure nested frontend/pages architecture is correctly scaffolded
    os.makedirs(os.path.dirname(SITEMAP_OUTPUT_PATH), exist_ok=True)
    with open(SITEMAP_OUTPUT_PATH, "w", encoding="utf-8") as f:
        f.write(xml_content)

    logger.info(f"Sitemap explicitly generated to: {SITEMAP_OUTPUT_PATH}")


if __name__ == "__main__":
    generate_sitemap()

File: api/scripts/generate-sitemap.js
#!/usr/bin/env node
// generate-sitemap.js — database-driven sitemap.xml generator.
//
// Queries every published content table for its slug + modification date,
// builds absolute https://www.thejesuswebsite.org/… URLs for all list and
// detail pages, and writes frontend/sitemap.xml. Uses only Node built-ins
// plus the existing better-sqlite3 config.
//
// Run:  npm run sitemap   or   node scripts/generate-sitemap.js
//
// This script is the source of truth for the sitemap. After this runs,
// the hand-maintained sitemap is replaced. robots.txt already points
// crawlers at /sitemap.xml.

const fs = require("fs");
const path = require("path");
const db = require("../config");

const BASE_URL = "https://thejesuswebsite.org";
const OUTPUT_PATH = path.resolve(
  __dirname,
  "..",
  "..",
  "frontend",
  "sitemap.xml",
);

/**
 * Escape XML special characters in text content.
 */
function xmlEscape(str) {
  if (typeof str !== "string") return "";
  return str
    .replace(/&/g, "&amp;")
    .replace(/</g, "&lt;")
    .replace(/>/g, "&gt;")
    .replace(/"/g, "&quot;")
    .replace(/'/g, "&apos;");
}

/**
 * Format a Date or date-like value as YYYY-MM-DD for sitemap <lastmod>.
 * Returns null if the value is not a valid date.
 */
function formatDate(value) {
  if (!value) return null;
  const d = new Date(value);
  if (isNaN(d.getTime())) return null;
  return d.toISOString().slice(0, 10);
}

// ── Section / index pages (static, no DB query) ──────────────────────────────

const SECTION_PAGES = [
  { loc: "/", priority: "1.0", changefreq: "weekly" },
  { loc: "/evidence/", priority: "0.9", changefreq: "weekly" },
  { loc: "/evidence/search.html", priority: "0.8", changefreq: "weekly" },
  { loc: "/evidence/arbor.html", priority: "0.7", changefreq: "monthly" },
  { loc: "/evidence/timeline/", priority: "0.7", changefreq: "monthly" },
  { loc: "/evidence/maps/", priority: "0.7", changefreq: "monthly" },
  {
    loc: "/evidence/maps/roman-empire.html",
    priority: "0.7",
    changefreq: "monthly",
  },
  { loc: "/evidence/maps/levant.html", priority: "0.7", changefreq: "monthly" },
  { loc: "/evidence/maps/judea.html", priority: "0.7", changefreq: "monthly" },
  {
    loc: "/evidence/maps/galilee.html",
    priority: "0.7",
    changefreq: "monthly",
  },
  {
    loc: "/evidence/maps/jerusalem.html",
    priority: "0.7",
    changefreq: "monthly",
  },
  {
    loc: "/evidence/maps/roman-empire/zoom-roman-empire.html",
    priority: "0.6",
    changefreq: "monthly",
  },
  {
    loc: "/evidence/maps/levant/zoom-levant.html",
    priority: "0.6",
    changefreq: "monthly",
  },
  {
    loc: "/evidence/maps/judea/zoom-judea.html",
    priority: "0.6",
    changefreq: "monthly",
  },
  {
    loc: "/evidence/maps/galilee/zoom-galilee.html",
    priority: "0.6",
    changefreq: "monthly",
  },
  {
    loc: "/evidence/maps/jerusalem/zoom-jerusalem.html",
    priority: "0.6",
    changefreq: "monthly",
  },
  { loc: "/contextual-essays/", priority: "0.7", changefreq: "monthly" },
  { loc: "/debate/", priority: "0.8", changefreq: "monthly" },
  { loc: "/news-and-blog/", priority: "0.8", changefreq: "weekly" },
  { loc: "/news-and-blog/blog/", priority: "0.7", changefreq: "weekly" },
  { loc: "/news-and-blog/news/", priority: "0.7", changefreq: "weekly" },
  { loc: "/resources/", priority: "0.6", changefreq: "monthly" },
  {
    loc: "/resources/sermons-and-sayings.html",
    priority: "0.5",
    changefreq: "monthly",
  },
  { loc: "/resources/parables.html", priority: "0.5", changefreq: "monthly" },
  { loc: "/resources/objects.html", priority: "0.5", changefreq: "monthly" },
  { loc: "/resources/people.html", priority: "0.5", changefreq: "monthly" },
  { loc: "/resources/sites.html", priority: "0.5", changefreq: "monthly" },
  { loc: "/resources/ot-verses.html", priority: "0.5", changefreq: "monthly" },
  {
    loc: "/resources/internal-witnesses.html",
    priority: "0.5",
    changefreq: "monthly",
  },
  {
    loc: "/resources/external-witnesses.html",
    priority: "0.5",
    changefreq: "monthly",
  },
  { loc: "/resources/places.html", priority: "0.5", changefreq: "monthly" },
  {
    loc: "/resources/world-events.html",
    priority: "0.5",
    changefreq: "monthly",
  },
  { loc: "/resources/miracles.html", priority: "0.5", changefreq: "monthly" },
  { loc: "/resources/events.html", priority: "0.5", changefreq: "monthly" },
  {
    loc: "/resources/apologetics.html",
    priority: "0.5",
    changefreq: "monthly",
  },
  {
    loc: "/resources/manuscripts.html",
    priority: "0.5",
    changefreq: "monthly",
  },
  { loc: "/resources/sources.html", priority: "0.5", changefreq: "monthly" },
  { loc: "/about.html", priority: "0.6", changefreq: "monthly" },
];

// ── Content type → URL mapping ───────────────────────────────────────────────
//
// URL patterns MUST match the frontend router (getResultUrl in search.js,
// getSegment-based URL parsing in each detail module).

const CONTENT_TYPES = [
  {
    table: "evidence",
    urlPrefix: "/evidence/single/",
    slugColumn: "slug",
    dateColumn: "updated_at",
    priority: "0.8",
  },
  {
    table: "context_essays",
    urlPrefix: "/contextual-essays/",
    slugColumn: "slug",
    dateColumn: "updated_at",
    priority: "0.7",
  },
  {
    table: "responses",
    urlPrefix: "/debate/responses/",
    slugColumn: "slug",
    dateColumn: "updated_at",
    priority: "0.7",
  },
  {
    table: "blog_posts",
    urlPrefix: "/news-and-blog/blog/",
    slugColumn: "slug",
    dateColumn: "updated_at",
    priority: "0.7",
  },
  {
    table: "historiography",
    urlPrefix: "/debate/historiography/",
    slugColumn: "slug",
    dateColumn: "updated_at",
    priority: "0.7",
  },
  {
    table: "challenges",
    urlPrefix: "/debate/popular-challenges/",
    slugColumn: "slug",
    dateColumn: null,
    priority: "0.7",
    extraWhere: "AND academic_popular = 'popular'",
  },
  {
    table: "challenges",
    urlPrefix: "/debate/academic-challenges/",
    slugColumn: "slug",
    dateColumn: null,
    priority: "0.7",
    extraWhere: "AND academic_popular = 'academic'",
  },
  {
    table: "news_articles",
    urlPrefix: "/news-and-blog/news/",
    slugColumn: "slug",
    dateColumn: null,
    priority: "0.7",
  },
];

// ── Build the sitemap ────────────────────────────────────────────────────────

function buildSitemap() {
  const urls = [];
  const today = new Date().toISOString().slice(0, 10);

  // 1. Section / index pages — use today as lastmod (they change with deploys).
  for (const page of SECTION_PAGES) {
    urls.push({
      loc: `${BASE_URL}${page.loc}`,
      lastmod: today,
      changefreq: page.changefreq,
      priority: page.priority,
    });
  }

  // 2. Detail pages from each content type.
  for (const type of CONTENT_TYPES) {
    const dateSelect = type.dateColumn
      ? `${type.dateColumn} AS mod_date`
      : "NULL AS mod_date";

    const rows = db
      .prepare(
        `SELECT ${type.slugColumn} AS slug, ${dateSelect}
         FROM ${type.table}
         WHERE published_draft = 1
           AND ${type.slugColumn} IS NOT NULL
           ${type.extraWhere || ""}`,
      )
      .all();

    for (const row of rows) {
      urls.push({
        loc: `${BASE_URL}${type.urlPrefix}${encodeURIComponent(row.slug)}`,
        lastmod: formatDate(row.mod_date) || today,
        priority: type.priority,
      });
    }
  }

  return urls;
}

function renderXml(urls) {
  let xml = '<?xml version="1.0" encoding="UTF-8"?>\n';
  xml += '<urlset xmlns="http://www.sitemaps.org/schemas/sitemap/0.9">\n';

  for (const url of urls) {
    xml += "  <url>\n";
    xml += `    <loc>${xmlEscape(url.loc)}</loc>\n`;
    xml += `    <lastmod>${url.lastmod}</lastmod>\n`;
    if (url.changefreq) {
      xml += `    <changefreq>${url.changefreq}</changefreq>\n`;
    }
    if (url.priority) {
      xml += `    <priority>${url.priority}</priority>\n`;
    }
    xml += "  </url>\n";
  }

  xml += "</urlset>\n";
  return xml;
}

// ── Main ─────────────────────────────────────────────────────────────────────

function main() {
  const dir = path.dirname(OUTPUT_PATH);
  if (!fs.existsSync(dir)) {
    console.error(`[sitemap] Output directory does not exist: ${dir}`);
    process.exit(1);
  }

  const urls = buildSitemap();
  const xml = renderXml(urls);

  fs.writeFileSync(OUTPUT_PATH, xml, "utf8");
  console.log(
    `[sitemap] Wrote ${urls.length} URLs to ${path.relative(process.cwd(), OUTPUT_PATH)}`,
  );
}

if (require.main === module) {
  main();
}

module.exports = { buildSitemap, renderXml, CONTENT_TYPES, SECTION_PAGES };
